### PR TITLE
Add logging of stdout to dbt graph run_command.

### LINKED ISSUE #1356: Print full log on dbt error
I am trying to debug cosmos, which is not successfully creating the DAG. Unfortunately, the error message i am getting is useless, just `using legacy validation callback`. Could [this line](https://github.com/astronomer/astronomer-cosmos/blob/06f242a3a6b1cc00781d169a38329f78509a122d/cosmos/dbt/graph.py#L155) be changed so that the error prints out stdout as well, so that i have more context to go on?

As an example, this is the full error:

```
[2024-12-02T15:49:40.651+0000] {graph.py:136} INFO - Running command: `/home/astro/.local/bin/dbt ls --output json --project-dir /tmp/tmpme078_tw --profiles-dir /tmp/cosmos/profile/3cc5d8bec2a8c876092d2b790d2cde171ff387a024cbaa5d615bf88275988ecd --profile derived_eng_dbt --target dev`
Traceback (most recent call last):
  File "/usr/local/airflow/dags/dbt/cosmos_dag.py", line 21, in <module>
    my_cosmos_dag = DbtDag(
                    ^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/cosmos/airflow/dag.py", line 26, in __init__
    DbtToAirflowConverter.__init__(self, *args, **specific_kwargs(**kwargs))
  File "/usr/local/lib/python3.12/site-packages/cosmos/converter.py", line 260, in __init__
    self.dbt_graph.load(method=render_config.load_method, execution_mode=execution_config.execution_mode)
  File "/usr/local/lib/python3.12/site-packages/cosmos/dbt/graph.py", line 402, in load
    self.load_via_dbt_ls()
  File "/usr/local/lib/python3.12/site-packages/cosmos/dbt/graph.py", line 461, in load_via_dbt_ls
    self.load_via_dbt_ls_without_cache()
  File "/usr/local/lib/python3.12/site-packages/cosmos/dbt/graph.py", line 580, in load_via_dbt_ls_without_cache
    nodes = self.run_dbt_ls(dbt_cmd, self.project_path, tmpdir_path, env)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/cosmos/dbt/graph.py", line 442, in run_dbt_ls
    stdout = run_command(ls_command, tmp_dir, env_vars)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/cosmos/dbt/graph.py", line 156, in run_command
    raise CosmosLoadDbtException(f"Unable to run {command} due to the error:\n{details}")
cosmos.dbt.graph.CosmosLoadDbtException: Unable to run ['/home/astro/.local/bin/dbt', 'ls', '--output', 'json', '--project-dir', '/tmp/tmpme078_tw', '--profiles-dir', '/tmp/cosmos/profile/3cc5d8bec2a8c876092d2b790d2cde171ff387a024cbaa5d615bf88275988ecd', '--profile', 'derived_eng_dbt', '--target', 'dev'] due to the error:
using legacy validation callback

```

If i edit the line referenced above to concat stdout + stderr i get
```
...

16:17:46  Running with dbt=1.8.7
16:17:47  Registered adapter: databricks=1.8.7
16:17:47  Unable to do partial parsing because config vars, config profile, or config target have changed
16:17:47  Unable to do partial parsing because profile has changed
16:17:48  Encountered an error:
Compilation Error
  Model 'model.derived_eng_dbt.testing_platform_test_results' (models/testing_platform/testing_platform_test_results.sql) depends on a source named 'silver_mysql_eng_testing.test_run' which was not found
```
which is exactly what i need.

### PR DESCRIPTION
**Pull request by KarolGongola**
## Description

Currently run_command functions put only stderr or stdout into logger. Which is misleading, because there are also important information in stdout even if something exists in stderr.

## Related Issue(s)

closes #1356

## Breaking Change?
NO

## Checklist

- [ ] I have made corresponding changes to the documentation (if required)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by tatiana — COMMENTED** [review]
Hi @KarolGongola, thank you very much for improving the logs

I'm happy with the improvement. Could you please fix the tests so they pass? It will also help us avoid any regressions in your log improvement.

Apparently, the following tests broke due to this change:
```
FAILED tests/dbt/test_graph.py::test_load_via_dbt_ls_with_non_zero_returncode - AssertionError: Regex pattern did not match.
 Regex: "Unable to run \\['.+dbt', 'deps', .*\\] due to the error:\\nSome stderr message"
 Input: "Unable to run ['/home/runner/.local/share/hatch/env/virtual/astronomer-cosmos/Za_bFbg4/tests.py3.11-2.9/bin/dbt', 'deps', '--project-dir', '/tmp/tmpivws28dg', '--profiles-dir', '/tmp/cosmos/profile/d426ea9bd2fe96408d8d2c4f150ab19011d7d7314fc17f613b902308e66fb2e7', '--profile', 'default', '--target', 'default'] due to the error:\nstderr: Some stderr message\nstdout: "
FAILED tests/dbt/test_graph.py::test_load_via_dbt_ls_with_runtime_error_in_stdout - AssertionError: Regex pattern did not match.
 Regex: "Unable to run \\['.+dbt', 'deps', .*\\] due to the error:\\nSome Runtime Error"
 Input: "Unable to run ['/home/runner/.local/share/hatch/env/virtual/astronomer-cosmos/Za_bFbg4/tests.py3.11-2.9/bin/dbt', 'deps', '--project-dir', '/tmp/tmp_jnwrslr', '--profiles-dir', '/tmp/cosmos/profile/d426ea9bd2fe96408d8d2c4f150ab19011d7d7314fc17f613b902308e66fb2e7', '--profile', 'default', '--target', 'default'] due to the error:\nstderr: \nstdout: Some Runtime Error"
= 2 failed, 66 passed, 5 skipped, 667 deselected, 2 warnings in 342.06s (0:05:42) =
```

**Comment by fjmacagno** [comment]
I wonder a little if it would be better to not split out the stderr and stdout in the first place, since it could be confusing. Still a substantial improvement.

**Comment by KarolGongola** [comment]
> Hi @KarolGongola, thank you very much for improving the logs
> 
> I'm happy with the improvement. Could you please fix the tests so they pass? It will also help us avoid any regressions in your log improvement.
> 
> Apparently, the following tests broke due to this change:
> 
> ```
> FAILED tests/dbt/test_graph.py::test_load_via_dbt_ls_with_non_zero_returncode - AssertionError: Regex pattern did not match.
>  Regex: "Unable to run \\['.+dbt', 'deps', .*\\] due to the error:\\nSome stderr message"
>  Input: "Unable to run ['/home/runner/.local/share/hatch/env/virtual/astronomer-cosmos/Za_bFbg4/tests.py3.11-2.9/bin/dbt', 'deps', '--project-dir', '/tmp/tmpivws28dg', '--profiles-dir', '/tmp/cosmos/profile/d426ea9bd2fe96408d8d2c4f150ab19011d7d7314fc17f613b902308e66fb2e7', '--profile', 'default', '--target', 'default'] due to the error:\nstderr: Some stderr message\nstdout: "
> FAILED tests/dbt/test_graph.py::test_load_via_dbt_ls_with_runtime_error_in_stdout - AssertionError: Regex pattern did not match.
>  Regex: "Unable to run \\['.+dbt', 'deps', .*\\] due to the error:\\nSome Runtime Error"
>  Input: "Unable to run ['/home/runner/.local/share/hatch/env/virtual/astronomer-cosmos/Za_bFbg4/tests.py3.11-2.9/bin/dbt', 'deps', '--project-dir', '/tmp/tmp_jnwrslr', '--profiles-dir', '/tmp/cosmos/profile/d426ea9bd2fe96408d8d2c4f150ab19011d7d7314fc17f613b902308e66fb2e7', '--profile', 'default', '--target', 'default'] due to the error:\nstderr: \nstdout: Some Runtime Error"
> = 2 failed, 66 passed, 5 skipped, 667 deselected, 2 warnings in 342.06s (0:05:42) =
> ```

sure, sorry for inconvenience,

**Comment by KarolGongola** [comment]
> I wonder a little if it would be better to not split out the stderr and stdout in the first place, since it could be confusing. Still a substantial improvement.

I think that just concatenating them without any info could be even more misleading because they wont be in chronological order.

**Comment by KarolGongola** [comment]
@tatiana Is it fine now?

**Comment by fjmacagno** [comment]
I meant keep them in chronological order by using the `stderr=STDOUT` param. https://docs.python.org/3/library/subprocess.html#subprocess.STDOUT

**Comment by KarolGongola** [comment]
> I meant keep them in chronological order by using the `stderr=STDOUT` param. https://docs.python.org/3/library/subprocess.html#subprocess.STDOUT

Ah, I see. It seems like good idea. Let me check how much changes it requires in the code.

**Comment by KarolGongola** [comment]
It is much better. Thanks @fjmacagno

File: cosmos/dbt/graph.py [hunk 1]
```diff
@@ -166,7 +166,7 @@ def run_command(command: list[str], tmp_dir: Path, env_vars: dict[str, str]) ->
     process = Popen(
         command,
         stdout=PIPE,
-        stderr=PIPE,
+        stderr=STDOUT,
```
Comment (tatiana) [inline cosmos/dbt/graph.py:169]: I can see a few downsides to this implementation:
1. Loss of stream separation: it is not possible to distinguish between regular output and error messages
2. Unnecessary memory overhead: combining stdout and stderr, if in large volume, can lead to memory overhead
3. Error handling: if we want to handle stderr and stdout logs differently, we cannot

Ordering chronologically the dbt logs between stdout and stderr does not justify these downsides.

I was okay with the original implementation - and once the tests are updated, the CI checks pass, I can approve and merge this change.

Comment (KarolGongola) [reply cosmos/dbt/graph.py:169]: Got it.

**Comment by KarolGongola** [comment]
@tatiana Is this fine for you now? Or do you prefer concatenation without "stdout" and "stderr" keywords?

**Review by tatiana — APPROVED** [review]
Thanks for the contribution and addressing the feedback, @KarolGongola , we're releasing this as part of 1.8

**Pull request merged**